core/state/statedb.go

```diff
@@ -635,39 +672,269 @@ func (s *StateDB) SubRefund(gas uint64) {
+					BalanceChangeReferred: make(map[common.Address]*big.Int, defaultNumOfSlots),
+					CodeChangeReferred:    make(map[common.Address]struct{}, defaultNumOfSlots),
+					AddrStateReferred:     make(map[common.Address]struct{}), // the address should be exist,
+					KVChangeReferred:      make(map[common.Address]Storage),  // the KV pair should be exist
+				}
+			}
+			if obj, exist := db.parallel.dirtiedStateObjectsInSlot[addr]; exist {
+				if _, exist := db.parallel.addrStateChangesInSlot[addr]; exist {
+					log.Info("Get balance from unconfirmed, AddrStateChange",
+						"my txIndex", s.txIndex, "DB's txIndex", i, "addr", addr)
+					s.parallel.unconfirmedRefList[db.txIndex].BalanceChangeReferred[addr] = new(big.Int).Set(obj.Balance())
+					return obj.Balance()
+				}
+				if _, exist := db.parallel.balanceChangesInSlot[addr]; exist { // only changed balance is reliable
+					log.Info("Get balance from unconfirmed, BalanceChange",
+						"my txIndex", s.txIndex, "DB's txIndex", i, "addr", addr)
+					s.parallel.unconfirmedRefList[db.txIndex].BalanceChangeReferred[addr] = new(big.Int).Set(obj.Balance())
+					return obj.Balance()
+				}
+
+			}
+		}
+	}
+	return nil
+}
+
+// for code, codeHash and codeSize, so we return the *StateObject
+func (s *StateDB) getCodeFromUnconfirmedDB(addr common.Address) *StateObject {
+	if addr == s.parallel.systemAddress {
+		// never get systemaddress from unconfirmed DB
+		return nil
+	}
+
+	// check the unconfirmed DB with range:  baseTxIndex -> txIndex -1(previous tx)
+	for i := s.txIndex - 1; i > s.parallel.baseTxIndex; i-- {
+		if db, ok := s.parallel.unconfirmedDBInShot[i]; ok {
+			if _, ok := s.parallel.unconfirmedRefList[db.txIndex]; !ok {
+				s.parallel.unconfirmedRefList[db.txIndex] = UnconfirmedReferDetail{
+					BalanceChangeReferred: make(map[common.Address]*big.Int, defaultNumOfSlots),
+					CodeChangeReferred:    make(map[common.Address]struct{}, defaultNumOfSlots),
+					AddrStateReferred:     make(map[common.Address]struct{}),
+					KVChangeReferred:      make(map[common.Address]Storage),
+				}
+			}
+			if obj, ok := db.parallel.dirtiedStateObjectsInSlot[addr]; ok {
+				if _, exist := db.parallel.addrStateChangesInSlot[addr]; exist {
+					log.Info("Get Code from unconfirmed, AddrStateChange",
+						"my txIndex", s.txIndex, "DB's txIndex", i, "addr", addr)
+					s.parallel.unconfirmedRefList[db.txIndex].CodeChangeReferred[addr] = struct{}{}
+					return obj
+				}
+				if _, exist := db.parallel.codeChangesInSlot[addr]; exist {
+					log.Info("Get Code FromUnconfirmedDB, CodeChange",
+						"my txIndex", s.txIndex, "DB's txIndex", i, "addr", addr)
+					s.parallel.unconfirmedRefList[db.txIndex].CodeChangeReferred[addr] = struct{}{}
+					return obj
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// for addr state check of: Exist(), Empty() and HasSuicided()
+func (s *StateDB) getAddrStateFromUnconfirmedDB(addr common.Address) *StateObject {
```

Comment:
is there any difference between `getStateObjectFromUnconfirmedDB` and `getAddrStateFromUnconfirmedDB`